boot/freeldr/bootsect/ntfs.S

```diff
@@ -0,0 +1,804 @@
+// SI points to the next run
+UnpackRun:
+    push bp
+    mov  bp, sp
+    sub  sp, HEX(10)
+    push ebx
+
+    // Unpack run header
+    mov  bl, byte ptr es:[si]
+    inc  si
+
+    mov  bh, bl
+    shr  bh, 4 
+    and  bl, 7 
+    mov  byte ptr [bp-2], bh     // BH contains the LCN length
+    mov  byte ptr [bp-1], bl     // BL contains the number of cluster length
+    
+    mov  al, bl
+    call UnpackLen
+    mov  dword ptr [bp - 8], ebx
+
+    mov  al, byte ptr [bp-2]
+    call UnpackLen
+    mov  cl, byte ptr es:[si-1]  // Fixup sign if last byte is > 255
+    test cl, HEX(80)
+    jz   NoSign
+    not  eax
+    add  ebx, eax
+
+    NoSign:
+    mov  eax, ebx
+    mov  ecx, dword ptr [bp - 8]
+    pop  ebx
+    mov  sp, bp
+    pop  bp
+    ret
+
+
+// Auxiliary function that unpacks n bytes in the memory
+// AL: size to unpack (max 4 bytes)
+// Returns: eax: the mask used to unpack (for negative number fixup)
+// Returns: ebx: the unpacked number
+UnpackLen:
+    push  cx
+    movzx ax, al
+
+    // Read the whole DWORD and then compute a mask to remove
+    // unneeded bytes to get correct size
+    xor   ebx, ebx
+    mov   ebx, dword ptr es:[si]
+    add   si, ax
+
+    
+    mov   cl, al    // Compute mask (2^(8*len) - 1)
+    shl   cl, 3
+    xor   eax, eax
+    inc   eax
+    shl   eax, cl
+    dec   eax
+
+    and   ebx, eax // Apply mask
+    pop   cx
+    ret
+
+// ES:AX : Address of the data runs
+// FS:BX : Address to write to
+// ECX    : LCN to read
+ReadINDX:
+    push  es
+    push  bx
+    call  ReadNonResidentAttribute
+    test  ax, ax
+    jz    PrintFileSystemError
+    cmp   dword ptr fs:[0], HEX(58444e49)   // Check if valid INDX
+    jnz   PrintFileSystemError              // jump if not valid
+    pop   bx
+    
+    mov   ax, fs
+    mov   es, ax
+    call  ApplyFixups
+    pop   es
+    ret
+
+// Finds an attribute header
+// ES:AX : pointer to the MFT entry
+// EBX   : type to find
+// returns: offset to the attribute header or 0 if not found
+FindAttributeHdr:
+    push  cx
+    push  si
+    push  edx
+    mov   si, ax
+    mov   cx, word ptr es:[si+HEX(14)]  // Get offset attribute
+    add   si, cx
+    FindAttributeHdrLoop:
+        mov   edx, dword ptr es:[si]    // Get attribute type 
+        cmp   edx, ebx                
+        jz    AttrFound
+        cmp   edx, HEX(ffffffff)
+        jz    AttrNotFound
+        add   cx, word ptr es:[si + 4]  // Add size of the attribute
+        add   si, word ptr es:[si + 4] 
+        jmp   FindAttributeHdrLoop
+    
+    AttrNotFound:
+    // Attribute not found, reset the offset
+    xor   cx, cx
+    AttrFound:
+    mov   ax, cx
+    pop   edx
+    pop   si
+    pop   cx
+    ret
+
+PrintFreeldrError:
+    mov  si, offset msgFreeldr
+    call PutChars
+    jmp  Reboot
+
+FreeLdr:
+    .word 'F', 'R', 'E', 'E', 'L', 'D', 'R', '.', 'S', 'Y', 'S'
+msgFreeldr:
+    .ascii "FreeLdr not found, cannot boot", CR, LF, NUL
+
```

Comment:
Maybe add and use a "Loading FreeLdr..." message as we do with our other boot sectors, see example here:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/bootsect/fat32.S;h=ff9c40af1211a7f7112b0a74d2bc7cdd876a8a0d;hb=HEAD#l530